Add live video stream support to ring-adapter

dickydoouk opened the pull request:
Uses streamVideo function of ring-client-api to start a live HLS stream.

This currently doesn't work using the default video viewer in WebThings due to limitations in the HLS library used.

Have opened this PR to get early feedback.

Commit:
Add video stream support to ring-adapter

Commit:
Add video stream support to ring-adapter

pdickinson1982 committed:
Apply linting and tidy code

pdickinson1982 committed:
Apply linting and tidy code

damooooooooooh commented:
Hi there, does this add a video stream property?  Does it show the last video from the device?  Thanks!

dickydoouk commented:
It adds a streaming property and an option to view the current live feed of the camera.

The streaming property must be set (Turned on) for the feed to be active.

![image](https://user-images.githubusercontent.com/33864869/111223906-aa805f00-85d5-11eb-89b8-bb6f14a16e29.png)

But currently I get an error when viewing the stream. Caused by Shaka player not supporting m3u8. Im looking into creating a PR to switch to HLS.js but struggling to get that to work locally atm.

damooooooooooh commented:
Hi, did you fix the issue related to this?

dickydoouk commented:
Still working on the viewing of the video feed from within webthings @damooooooooooh albeit slowly. Will keep this PR and you posted with progress?

damooooooooooh commented:
No worries!

Commit:
Switch to dash+xml format as this has better shaka player support

dickydoouk commented:
<img width="605" alt="Screenshot 2021-08-05 at 00 29 44" src="https://user-images.githubusercontent.com/33864869/128268451-f48c4053-6ac5-46a6-9482-ed7b8c475aea.png">

Evidence of it working in Chrome

damooooooooooh commented:
Cool, does this display the last motion of the device?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: dickydoouk ***@***.***>
Sent: Thursday, August 5, 2021 12:30:56 AM
To: damooooooooooh/ring-adapter ***@***.***>
Cc: damooooooooooh ***@***.***>; Mention ***@***.***>
Subject: Re: [damooooooooooh/ring-adapter] Add live video stream support to ring-adapter (#5)


[Screenshot 2021-08-05 at 00 29 44]<https://user-images.githubusercontent.com/33864869/128268451-f48c4053-6ac5-46a6-9482-ed7b8c475aea.png>

Evidence of it working in Chrome

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/damooooooooooh/ring-adapter/pull/5#issuecomment-893042852>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAKDYDXFBBMDOMSFT5RUMLLT3HETBANCNFSM4ZHI5NOA>.

damooooooooooh commented:
Hi this doesnt work for me.  Error: Ffmpeg is not installed.  How do we bundle the dependency?

Also, for my Floodlight camera, the streaming properties are not generated as the data.features object doesn't have the show_vod_settings property......

2021-08-05 13:57:10.408 INFO   : ring-adapter: { motion_zone_recommendation: false,
2021-08-05 13:57:10.409 INFO   : ring-adapter:   motions_enabled: true,
2021-08-05 13:57:10.410 INFO   : ring-adapter:   show_recordings: true,
2021-08-05 13:57:10.411 INFO   : ring-adapter:   advanced_motion_enabled: false,
2021-08-05 13:57:10.412 INFO   : ring-adapter:   people_only_enabled: false,
2021-08-05 13:57:10.413 INFO   : ring-adapter:   shadow_correction_enabled: false,
2021-08-05 13:57:10.415 INFO   : ring-adapter:   motion_message_enabled: false,
2021-08-05 13:57:10.416 INFO   : ring-adapter:   night_vision_enabled: false,
2021-08-05 13:57:10.417 INFO   : ring-adapter:   rich_notifications_eligible: true,
2021-08-05 13:57:10.418 INFO   : ring-adapter:   show_24x7_lite: true,
2021-08-05 13:57:10.419 INFO   : ring-adapter:   show_offline_motion_events: false,
2021-08-05 13:57:10.420 INFO   : ring-adapter:   cfes_eligible: null,
2021-08-05 13:57:10.421 INFO   : ring-adapter:   sheila_camera_eligible: null,
2021-08-05 13:57:10.422 INFO   : ring-adapter:   sheila_camera_processing_eligible: null }

dickydoouk commented:
I have tested this both on my ubuntu dev machine and my raspberry pi running webthings and it does view the camera feed these probably have ffmpeg pre-installed. As the ring-adapted depends on ring-client-api it should download and install ffmpeg-for-homebridge. I get the following message when running ```npm install``` which webthings should do on add-on installation:

```
ffmpeg has been downloaded to /home/pi/.webthings/addons/ring-adapter/node_modules/ffmpeg-for-homebridge/ffmpeg
Thank you for using https://github.com/homebridge/ffmpeg-for-homebridge
```

With regards the other issue: I only have the ring doorbell and don't own a ring floodlight camera so won't be able to fix that without some assistance. how did you list the ring-adapter properties. Maybe there is a better property we can use than show_vod_settings e.g. show_recordings

damooooooooooh commented:
What platform do you build on?

dickydoouk commented:
I build on an Ubuntu Linux dist and also have a raspberry pi with stock webthings image and checkout the ring addon directly from my fork. I would imagine if you are building on a mac/windows machine that ffmpeg is not installed by default.

Commit:
Ensure streaming property is updated when stream auto turns off after 5 mins

Commit:
Bump version no

damooooooooooh commented:
Hey, I just pulled this and tested after building on a mac.  I still get the below error 

Error: Ffmpeg is not installed.  See https://github.com/dgreif/ring/wiki/FFmpeg for directions.
ERROR  : ring-adapter:     at RingCamera.<anonymous> (/home/pi/.webthings/addons/ring-adapter/node_modules/ring-client-api/lib/api/ring-camera.js:451:23)

I see the ffmpeg binary at /home/pi/.webthings/addons/ring-adapter/node_modules/ffmpeg-for-homebridge

Any ideas?

damooooooooooh commented:
Got this working, just replaced the ffmpeg binary with raspbian one.  Quick question, will this stream indefinitely and consume disk space?

Also, for some reason, my ring floodlight doesnt return hasCamera as true.

ring-device.js --> Is this required?
this.hasCamera = camera.data.features &&
      camera.data.features.show_vod_settings;

dickydoouk commented:
Glad you got it working. yes the stream will auto stop after 5 minutes to conserve data & battery. So isn't that useful to view the stream over longer periods of time. But this could be made configurable in the future for those who want it.

As for the ring floodlight unfortunately I do not possess one so am unable to test the properties of this device with the camera but hopefully its interface is similar to that of the doorbells.

damooooooooooh committed:
Fix hasCamera Property

Use data.settings.enable_vod.

damooooooooooh committed:
Fix media Path

Remove reference to Mozilla IOT platfrom.

dickydoouk commented:
Is there anything else that prevents this PR from being merged to master?

damooooooooooh commented:
No, ive tested it for a few weeks and all good, Ill get things updated on the next day.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: ***@***.***>
Sent: 27 February 2022 16:57
To: ***@***.***>
Cc: ***@***.***>; ***@***.***>
Subject: Re: [damooooooooooh/ring-adapter] Add live video stream support to ring-adapter (#5)


Is there anything else that prevents this PR from being merged to master?

—
Reply to this email directly, view it on GitHub<https://github.com/damooooooooooh/ring-adapter/pull/5#issuecomment-1053614789>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAKDYDRMCRKD4IDMPHWBB4DU5JJZDANCNFSM4ZHI5NOA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

dickydoouk commented:
Any chance this could be merged yet? I don't have write access.

damooooooooooh commented:
Hey, looking at this today.  Having issues with ffmpeg when I install on pi.  I have to take manual actions, to update the ffmpeg version.  Will need to create two builds.  Will hopefully do it today.

damooooooooooh commented:
We need to package this in accordance with the guidelines here with [binaries](https://github.com/WebThingsIO/addon-list#packages-with-binaries).  What platform are you running your version on? linux x64?  Im running on arm so the binary differs.

The main issue here is around dependencies, ring-api-client has the ffmpeg-for-homebridge as a dependency which in turn has an install script that detects the build platform and includes the relevant ffmpeg binary.  This is no good as webthings packages  use the sha256 signature so package cannot be modified post install.  We need to create a build for each platform supported and build separately

damooooooooooh committed:
Proposed changes for Release

Updated build script to handle multiple build for ffmpeg.  Altered how the stream property and video property are handled.

damooooooooooh commented:
Hi, can you pull my updates and build locally (package.sh).  Will now produce linux-x64 package that you should install and test.  Please check how the stream property now behaves e.g. the Video property is removed if the stream property is false.

Commit:
bump ring-client-api version

dickydoouk commented:
Appreciate this is nearly 2yrs ago but I've rebased this pull-request and all seems working on my setup.

damooooooooooh commented:
Hey I'm happy to make you a co-owner as I don't use the platform anymore.

dickydoouk commented:
Thanks much appreciated

Commit:
Add record on motion/ding capability

dickydoouk commented:
Thanks. Please let me know when you have enabled write access to the repo
for me as I currently can't merge the PR.

On Wed, 6 Mar 2024 at 23:41, damooooooooooh ***@***.***>
wrote:

> Hey I'm happy to make you a co-owner as I don't use the platform anymore.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/damooooooooooh/ring-adapter/pull/5#issuecomment-1982050634>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AICLZJO254KXUWHICEFXEFTYW6SSPAVCNFSM4ZHI5NOKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCOJYGIYDKMBWGM2A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

damooooooooooh commented:
Hi what platform are you running on? The lastest ring-api-client (12.1.0) on my side is failing to load, potentially related to ECMAScript2020 on the webthings image.  Did you having similar issues?

/home/node/.webthings/addons/ring-adapter/node_modules/ring-client-api/lib/api.js:131
ERROR  : ring-adapter:             devicesById[id]?.processPushNotification(notification);

SyntaxError: Unexpected token '.'

dickydoouk commented:
Yes I downgraded the client in the end as it requires a newer version of
Node to run. I believe Optional chaining (?.) is available from Node.js 14
but raspberry pi webthings image uses Node.js 10 which does not support
this hence the error. I've pinned the library to use ring client 9.18.6 in
the PR. Everything seems to work fine.

I also added the ability to enable automatic recording of dings & motions.
Although there is currently no facility in webthings to view them (Yet) but
they are stored in .webthings/media/ring-doorbell/recordings and
automatically delete after a configurable number of recordings.

On Fri, 8 Mar 2024, 14:43 damooooooooooh, ***@***.***> wrote:

> Hi what platform are you running on? The lastest ring-api-client (12.1.0)
> on my side is failing to load, potentially related to ECMAScript2020 on the
> webthings image. Did you having similar issues?
>
>
> /home/node/.webthings/addons/ring-adapter/node_modules/ring-client-api/lib/api.js:131
> ERROR : ring-adapter:
> devicesById[id]?.processPushNotification(notification);
>
> SyntaxError: Unexpected token '.'
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/damooooooooooh/ring-adapter/pull/5#issuecomment-1985816364>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AICLZJIALAJPTH3BUW33BHTYXHFBVAVCNFSM4ZHI5NOKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCOJYGU4DCNRTGY2A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

damooooooooooh commented:
I will be issuing a pull request to you PR in the next day.  I installed this and noticed few issues with a few of my Ring devices.
